onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu

```diff
@@ -192,36 +194,48 @@
 
-#define CASE_NEXT_SIZE(next_size_value)               \
-  case next_size_value: {                             \
-    if (flag_vec4) {                                  \
-      constexpr int block_size = next_size_value / 4; \
-      LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(4);         \
-    } else if (flag_vec2) {                           \
-      constexpr int block_size = next_size_value / 2; \
-      LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(2);         \
-    } else {                                          \
-      if (next_size_value <= kMaxBlockSize) {         \
-        constexpr int block_size = next_size_value;   \
-        LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(1);       \
-      } else {                                        \
-        LAUNCH_SKIP_LAYER_NORM_KERNEL();              \
-      }                                               \
-    }                                                 \
+#define CASE_NEXT_SIZE(next_size_value)                                       \
+  case next_size_value: {                                                     \
+    static_assert(next_size_value > kSizes[0] && next_size_value < kMaxSize); \
+    if (flag_vec4) {                                                          \
+      constexpr int block_size = next_size_value / 4;                         \
+      LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(4);                                 \
+    } else if (flag_vec2) {                                                   \
+      constexpr int block_size = next_size_value / 2;                         \
+      LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(2);                                 \
+    } else {                                                                  \
+      if (next_size_value <= kMaxBlockSize) {                                 \
+        constexpr int block_size = next_size_value;                           \
+        LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(1);                               \
+      } else {                                                                \
+        constexpr int block_size = 256;                                       \
+        LAUNCH_SKIP_LAYER_NORM_KERNEL();                                      \
+      }                                                                       \
+    }                                                                         \
   } break
 
   switch (next_size) {
-    CASE_NEXT_SIZE(kSizes[0]);
+    case kSizes[0]: {
+      constexpr int block_size = kSizes[0];
+      // TODO: Add back the small TensorRT kernel for 32. No need to use vertorized kernel for such small size.
+      LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(1);
+      break;
+    }
     CASE_NEXT_SIZE(kSizes[1]);
     CASE_NEXT_SIZE(kSizes[2]);
     CASE_NEXT_SIZE(kSizes[3]);
     CASE_NEXT_SIZE(kSizes[4]);
     CASE_NEXT_SIZE(kSizes[5]);
-    CASE_NEXT_SIZE(kSizes[6]);
+    // kMaxSize shall not run vectorized kernel since ld might be larger than kMaxSize.
+    default: {
```

Comment:
it is already handled in CanVectorized

Reply:
Right. CanVectorized will always returns false for kMaxSize. The underlying reason is limit of vectorized kernel implementation, so I add the comment here.

For example, ld=4096, next_size=2048. If vectorizied is used, there will be 512 threads, each thread reads 4. Then the kernel only read 2048 instead of 4096 elements per row.